Nondeterminism observed on pandas in typeshed primer

mypy_primer in typeshed has started flakily showing the introduction or removal of an error on pandas.

This typically looks like:
```
pandas (https://github.com/pandas-dev/pandas)
+ pandas/core/computation/ops.py:328: error: Need type annotation for "_binary_ops_dict" (hint: "_binary_ops_dict: dict[<type>, <type>] = ...")  [var-annotated]
```

You can see this in a number of typeshed PRs https://github.com/python/typeshed/pulls?q=is%3Apr+_binary_ops_dict

Here's an example error from April 7 https://github.com/python/typeshed/pull/15622#issuecomment-4197149226. Typeshed updated to 1.20 on April 1.

The interesting thing is I suspect this will not repro with stock mypy (e.g. we haven't seen it in mypy's mypy_primer)

I think what is going in is that you also need these typeshed PRs to `operator.py`. I reverted these PRs for us because they are bad for our users (for other reasons): https://github.com/python/mypy/pull/21232#issuecomment-4374615030

I set Codex on this a while ago but didn't submit its PR yet because I didn't feel like it was addressing the root cause. Codex thought it was because operator and _operator are part of an SCC and some types are different dependent on the order in which the files in the SCC are processed.

Depending on the order, some dict in pandas gets inferred as either `dict[str, Overload(...)]` or `dict[str, function]`, and one of the two leads to errors. Codex proposed to fix that by changing how we union those two types, which felt like a bandaid.

If you want the LLM's full verbose output:

<details>

**Title**

Nondeterministic inference for `operator` functions can make `dict(zip(...))` alternate between precise callable types and `builtins.function`

**Summary**

I observed nondeterministic mypy output in mypy-primer on pandas. The reported diff was that this pandas error sometimes appeared and sometimes disappeared with identical code:

```text
pandas/core/computation/ops.py:328: error: Need type annotation for "_binary_ops_dict"  [var-annotated]
```

The code is roughly:

```python
import operator

ARITH_OPS_SYMS = ("+", "-", "*", "/", "**", "//", "%")
_arith_ops_funcs = (
    operator.add,
    operator.sub,
    operator.mul,
    operator.truediv,
    operator.pow,
    operator.floordiv,
    operator.mod,
)
_arith_ops_dict = dict(zip(ARITH_OPS_SYMS, _arith_ops_funcs, strict=True))

_binary_ops_dict = {}

for d in (_cmp_ops_dict, _bool_ops_dict, _arith_ops_dict):
    _binary_ops_dict.update(d)
```

Repeated runs against the same checkout sometimes infer `_arith_ops_dict` as a precise overloaded-function dict, and sometimes as a bare `builtins.function` dict. That then changes the loop item type for `d`.

**Observed shapes**

In one run:

```text
for d in (...): d
# dict[str, def (Any, Any) -> Any]
```

In another run:

```text
for d in (...): d
# dict[str, def (Any, Any) -> Any] | dict[str, function]
```

The latter shape prevents partial dict inference for `_binary_ops_dict = {}`, so mypy emits the annotation error.

**Likely source of nondeterminism**

The root appears to be interaction between typeshed’s `operator` / `_operator` stubs and mypy’s SCC processing.

In current typeshed, there is an import cycle:

```python
# operator.pyi
from _operator import add, sub, mul, ...

# _operator.pyi
from operator import attrgetter, itemgetter, methodcaller
```

So `operator` and `_operator` are in the same SCC. Mypy’s build graph / SCC code passes around sets in a few places, for example SCCs and ready groups. Depending on processing order within that SCC, `operator.add` and friends may be seen as precise overloaded functions or as the coarser fallback type `builtins.function`.

That gives `_arith_ops_dict` two possible inferred shapes:

```text
dict[str, Overload(...)]
```

or:

```text
dict[str, function]
```

Then later union simplification preserves this difference, which leaks into pandas’ `_binary_ops_dict` inference.

**Notes**

- This was observed with mypy-primer using compiled mypy.
- The pandas/typeshed code was unchanged between runs.
- The typeshed PR that exposed it was unrelated (`sys._jit`), so it likely just perturbed processing enough to reveal an existing nondeterminism.
- A local mitigation was to make union simplification prefer a precise function-like type over bare `builtins.function` when the latter is only a fallback inside the same container shape. That makes the pandas loop infer one deterministic type, but it does not fully explain/fix the earlier `operator`/`_operator` shape flip.

</details>

--- a/mypy/solve.py
+++ b/mypy/solve.py
@@ -17,6 +17,7 @@ from mypy.types import (
     AnyType,
     Instance,
     NoneType,
+    Overloaded,
     ParamSpecType,
     ProperType,
     TupleType,
@@ -253,6 +254,9 @@ def _join_sorted_key(t: Type) -> int:
         return -2
     if isinstance(t, NoneType):
         return -1
+
+    if isinstance(t, Overloaded):
+        return 1
     return 0
 
 

--- a/test-data/unit/check-generics.test
+++ b/test-data/unit/check-generics.test
@@ -3542,7 +3542,7 @@ reveal_type(C.foo)  # N: Revealed type is "def [T] (self: __main__.B[T`1]) -> T`
 reveal_type(C[int].foo)  # N: Revealed type is "def (self: __main__.B[builtins.int]) -> builtins.int"
 reveal_type(D.foo)  # N: Revealed type is "def (self: __main__.B[builtins.int]) -> builtins.int"
 
-[case testDeterminismFromJoinOrderingInSolver]
+[case testDeterminismFromJoinOrderingInSolver1]
 # Used to fail non-deterministically
 # https://github.com/python/mypy/issues/19121
 from __future__ import annotations
@@ -3594,6 +3594,46 @@ def draw_none(
         takes_int_str_none(c2)
         takes_int_str_none(c3)
 [builtins fixtures/tuple.pyi]
+
+[case testDeterminismFromJoinOrderingInSolver2]
+# Used to fail non-deterministically
+# https://github.com/python/mypy/issues/21445
+from typing import Generic, Iterable, TypeVar, overload
+
+class A: ...
+
+@overload
+def f0(a: A, b: object, /) -> object: ...
+@overload
+def f0(a: object, b: int, /) -> object: ...
+def f0(a, b, /): ...
+
+@overload
+def f1(a: int, b: object, /) -> object: ...
+@overload
+def f1(a: object, b: A, /) -> object: ...
+def f1(a, b, /): ...
+
+def g(a, b, /): ...
+
+T = TypeVar("T")
+K = TypeVar("K")
+V = TypeVar("V")
+
+class ziplike(Generic[T]):
+    def __new__(cls, x: str, y: tuple[V, ...], /) -> ziplike[tuple[str, V]]:
+        raise
+    def __iter__(self) -> ziplike[T]:
+        return self
+    def __next__(self) -> T:
+        raise
+
+class dictlike(Generic[K, V]):
+    def __init__(self, arg: Iterable[tuple[K, V]]) -> None: pass
+
+x = dictlike(ziplike("012", (f0, f1, g)))
+reveal_type(x)  # N: Revealed type is "__main__.dictlike[builtins.str, Overload(def (Any, Any) -> Any, def (Any, Any) -> Any, def (Any, Any) -> Any, def (Any, Any) -> Any)]"
+[builtins fixtures/dict.pyi]
 
 [case testPropertyWithGenericSetter]
 from typing import TypeVar

